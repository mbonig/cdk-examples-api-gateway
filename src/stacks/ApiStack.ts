import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Deployment, RestApi, Stage } from 'aws-cdk-lib/aws-apigateway';
import { Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
import { AttributeType, TableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { ARecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { ApiGateway } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

export interface ApiStackProps extends StackProps {
  hostedZoneId: string;
}

export class ApiStack extends Stack {
  api: RestApi;
  table: TableV2;
  private deployment!: Deployment;
  private apiDomainName: string;

  constructor(scope: Construct, id: string, private props: ApiStackProps) {
    super(scope, id, props);

    const apiDomainName = this.apiDomainName = 'api.example.matthewbonig.com';
    this.api = this.createApi();
    this.table = this.createTable();

    new CfnOutput(this, 'BaseApiGatewayURL', {
      value: `https://${this.api.restApiId}.execute-api.${this.region}.amazonaws.com/prod/`,
    });
    new CfnOutput(this, 'CustomURL', {
      value: `https://${apiDomainName}/v1/`,
    });

  }

  private createApi(): RestApi {
    const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: this.props.hostedZoneId,
      zoneName: 'example.matthewbonig.com',
    });
    const api = new RestApi(this, 'MyRestApi', {
      deploy: false,
      restApiName: 'CdkCourseApi',
      domainName: {
        domainName: this.apiDomainName,
        certificate: new Certificate(this, 'Certificate', {
          domainName: this.apiDomainName,
          validation: CertificateValidation.fromDns(hostedZone),
        }),
      },
    });

    new ARecord(this, 'ApiAliasRecord', {
      target: {
        aliasTarget: new ApiGateway(api),
      },
      recordName: 'api',
      zone: hostedZone,
    });

    api.root.addMethod('ANY');

    const deployment = this.deployment = new Deployment(this, 'Deployment', {
      api: api,
    });
    new Stage(this, 'Prod', {
      deployment,
      stageName: 'v1',
    });

    return api;
  }

  private createTable(): TableV2 {
    return new TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'pk',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });
  }

  public addToDeployment(resources: any[]) {
    for (const resource of resources) {
      this.deployment.addToLogicalId(resource);
    }
  }
}
