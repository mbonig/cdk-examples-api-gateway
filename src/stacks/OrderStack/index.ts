export * from './OrderStack';
export { OrderStackTests, OrderStackTestProps } from './OrderStackTests/OrderStackTests';

