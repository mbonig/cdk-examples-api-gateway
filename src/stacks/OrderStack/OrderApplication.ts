import { IOrderRepository } from './OrderRepository';
import { IdentifiableModel } from '../common/handlers/IdentifiableModel';

export interface OrderLineItem {
  productId: string;
  quantity: number;
  price: number;
}

export interface Order extends IdentifiableModel {
  personId: string;
  items: OrderLineItem[];
  orderTotal: number;
}


export interface IOrderApplication {
  getById(id: string): Promise<Order>;

  getAll(): Promise<Order[]>;
}

export class OrderApplication implements IOrderApplication {
  constructor(private orderRepository: IOrderRepository) {}

  getAll(): Promise<Order[]> {
    return this.orderRepository.getAll();
  }

  getById(id: string): Promise<Order> {
    return this.orderRepository.get(id);
  }

  async save(order: Partial<Order>) {
    this.validateOrder(order);
    return this.orderRepository.save({
      orderId: `order-${order.personId}-${new Date().toISOString()}`,
      personId: order.personId!,
      items: order.items!,
      orderTotal: order.orderTotal!,
    });
  }

  private validateOrder(order: Partial<Order>) {
    if (!order.personId) {
      throw new Error('Please provide a personId');
    }
    if (!order.items || order.items.length === 0) {
      throw new Error('Please provide at least one item');
    }
  }
}
