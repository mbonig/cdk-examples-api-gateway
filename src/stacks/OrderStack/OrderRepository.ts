import { Order } from './OrderApplication';
import { IDatabaseAdapter, TableModel } from '../common/IDatabaseAdapter';
import { IDatabaseRepository } from '../PersonStack/DatabaseRepository';

export interface IOrderRepository extends IDatabaseRepository<Order> {
}

export class OrderRepository implements IOrderRepository {
  constructor(private databaseAdapter: IDatabaseAdapter, private indexName: string) {

  }

  async get(id: string): Promise<Order> {
    const orderAndLineItems = await this.databaseAdapter.queryItem(id);
    return this.map(orderAndLineItems);
  }

  async getAll(): Promise<Order[]> {
    const ordersAndLines = await this.databaseAdapter.getAll(this.indexName);
    const orders = ordersAndLines.reduce((acc, item) => {
      if (!acc[item.pk]) {
        acc[item.pk] = [];
      }
      acc[item.pk].push(item);
      return acc;
    }, {} as { [key: string]: TableModel[] });
    return Object.values(orders).map(order => this.map(order));
  }

  async save(order: Order): Promise<Order> {
    await this.databaseAdapter.putItem({
      pk: order.orderId,
      sk: 'header',
      orderTotal: order.orderTotal,
      orderId: order.orderId,
      personId: order.personId,
    });
    let i = 1;
    for (const item of order.items) {
      await this.databaseAdapter.putItem({
        pk: order.orderId,
        sk: `line#${i.toString().padStart(3, '0')}`,
        orderId: order.orderId,
        ...item,
      });
      i++;
    }

    return order;
  }

  private map(orderAndLineItems: TableModel[]) {
    const header = orderAndLineItems.find(item => item.sk === 'header')!;
    const lineItems = orderAndLineItems.filter(item => item.sk !== 'header')!;

    return {
      orderId: header.pk,
      personId: header.personId,
      items: lineItems.map(item => ({
        productId: item.productId,
        quantity: item.quantity,
        price: item.price,
      })),
      orderTotal: header.orderTotal,
    };
  }
}
