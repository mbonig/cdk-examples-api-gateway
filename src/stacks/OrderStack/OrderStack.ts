import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { INDEX_NAME as GET_INDEX_NAME, TABLE_NAME as GET_TABLE_NAME } from './OrderStack.GetOrderHandler';
import { INDEX_NAME as POST_INDEX_NAME, TABLE_NAME as POST_TABLE_NAME } from './OrderStack.PostOrderHandler';
import { OrderStackTests } from './OrderStackTests/OrderStackTests';
import { RestApiResourceStack } from '../RestApiResourceStack';
import { RestResourceProps } from '../RestResourceStackProps';

interface OrderStackProps extends RestResourceProps {
}

export class OrderStack extends RestApiResourceStack {
  public deploymentDependencies: any[] = [];

  constructor(scope: Construct, id: string, props: OrderStackProps) {
    super(scope, id, props);
    const indexName = 'orders';

    const getHandler = new NodejsFunction(this, 'GetOrderHandler', {
      environment: {
        [GET_TABLE_NAME]: props.table.tableName,
        [GET_INDEX_NAME]: indexName,
      },
    });

    const postHandler = new NodejsFunction(this, 'PostOrderHandler', {
      environment: {
        [POST_TABLE_NAME]: props.table.tableName,
        [POST_INDEX_NAME]: indexName,
      },
    });
    props.table.grantReadWriteData(postHandler);
    props.table.grantReadData(getHandler);

    const orderResource = this.api.root.addResource('order', {});
    const getMethod = orderResource.addMethod('GET', new LambdaIntegration(getHandler));
    const postMethod = orderResource.addMethod('POST', new LambdaIntegration(postHandler));

    const getByIdMethod = orderResource
      .addResource('{id}', {})
      .addMethod('GET', new LambdaIntegration(getHandler));

    props.table.addGlobalSecondaryIndex({
      indexName: indexName,
      partitionKey: {
        name: 'orderId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });

    new OrderStackTests(this, 'Tests', {
      getOptions: {
        apiEndpoint: 'https://api.example.matthewbonig.com/v1/order/',
      },
      postOptions: {
        apiEndpoint: 'https://api.example.matthewbonig.com/v1/order/',
        table: props.table,
      },
    });

    this.deploymentDependencies.push(
      { method: getMethod.methodId },
      { method: getByIdMethod.methodId },
      { method: postMethod.methodId },
    );
  }
}
