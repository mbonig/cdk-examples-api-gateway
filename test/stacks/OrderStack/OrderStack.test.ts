import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Code, CodeConfig } from 'aws-cdk-lib/aws-lambda';
import { ApiStack } from '../../../src/stacks/ApiStack';
import { OrderStack } from '../../../src/stacks/OrderStack';


let fromAssetMock: jest.SpyInstance;
beforeAll(() => {
  fromAssetMock = jest.spyOn(Code, 'fromAsset').mockReturnValue({
    isInline: false,
    bind: (): CodeConfig => {
      return {
        s3Location: {
          bucketName: 'my-bucket',
          objectKey: 'my-key',
        },
      };
    },
    bindToResource: () => {
      return;
    },
  } as any);
});

afterAll(() => {
  fromAssetMock?.mockRestore();
});

test('Snapshot', () => {
  const app = new App();
  const { api, table } = new ApiStack(app, 'Api', {});
  const stack = new OrderStack(app, 'test', {
    api: api,
    table: table,
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
