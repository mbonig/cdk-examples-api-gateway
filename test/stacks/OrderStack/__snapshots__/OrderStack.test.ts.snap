// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": {
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9b642d28372ac0688d94d58f71770ceba997be43db88ea4b1cd0081a3ac5231e.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "TestsGetOrderTest88AC940D",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "TestsPostOrderTestEB3C878D",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetOrderHandler4576A72C": {
      "DependsOn": [
        "GetOrderHandlerServiceRoleDefaultPolicy56174588",
        "GetOrderHandlerServiceRoleCBF11364",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "my-bucket",
          "S3Key": "my-key",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "INDEX_NAME": "orders",
            "TABLE_NAME": {
              "Fn::ImportValue": "Api:ExportsOutputRefMyTable794EDED1558A264E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "GetOrderHandlerServiceRoleCBF11364",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetOrderHandlerServiceRoleCBF11364": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetOrderHandlerServiceRoleDefaultPolicy56174588": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetOrderHandlerServiceRoleDefaultPolicy56174588",
        "Roles": [
          {
            "Ref": "GetOrderHandlerServiceRoleCBF11364",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PostOrderHandlerE150467D": {
      "DependsOn": [
        "PostOrderHandlerServiceRoleDefaultPolicy0083823D",
        "PostOrderHandlerServiceRole9A777913",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "my-bucket",
          "S3Key": "my-key",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "INDEX_NAME": "orders",
            "TABLE_NAME": {
              "Fn::ImportValue": "Api:ExportsOutputRefMyTable794EDED1558A264E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "PostOrderHandlerServiceRole9A777913",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostOrderHandlerServiceRole9A777913": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostOrderHandlerServiceRoleDefaultPolicy0083823D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostOrderHandlerServiceRoleDefaultPolicy0083823D",
        "Roles": [
          {
            "Ref": "PostOrderHandlerServiceRole9A777913",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RestApiorder8B1A6481": {
      "Properties": {
        "ParentId": {
          "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyRestApi2D1F47A9RootResourceIdA8A99F67",
        },
        "PathPart": "order",
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiorderGETApiPermissionTesttestRestApi41EB688DGETorder6A2A42AE": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetOrderHandler4576A72C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/test-invoke-stage/GET/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderGETApiPermissiontestRestApi41EB688DGETorder843CE060": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetOrderHandler4576A72C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/*/GET/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderGETE90D3FAE": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "GetOrderHandler4576A72C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiorder8B1A6481",
        },
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiorderPOST6B9C6E9D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "PostOrderHandlerE150467D",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiorder8B1A6481",
        },
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiorderPOSTApiPermissionTesttestRestApi41EB688DPOSTorderC5FC9D25": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "PostOrderHandlerE150467D",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/test-invoke-stage/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderPOSTApiPermissiontestRestApi41EB688DPOSTorder4D010E41": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "PostOrderHandlerE150467D",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/*/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderidD816925B": {
      "Properties": {
        "ParentId": {
          "Ref": "RestApiorder8B1A6481",
        },
        "PathPart": "{id}",
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiorderidGETApiPermissionTesttestRestApi41EB688DGETorderid3351AEDB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetOrderHandler4576A72C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/test-invoke-stage/GET/order/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderidGETApiPermissiontestRestApi41EB688DGETorderid4CDAC4AB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetOrderHandler4576A72C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/*/GET/order/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiorderidGETF268739C": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "GetOrderHandler4576A72C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiorderidD816925B",
        },
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestsGetOrderTest88AC940D": {
      "DependsOn": [
        "TestsGetOrderTestServiceRoleDFB27BF0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "my-bucket",
          "S3Key": "my-key",
        },
        "Environment": {
          "Variables": {
            "API_ENDPOINT": "https://api.example.matthewbonig.com/v1/order/",
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestsGetOrderTestServiceRoleDFB27BF0",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestsGetOrderTestCurrentVersion2705E2D1c5f2213f45730c0d6e64049522f9d2f3": {
      "Properties": {
        "FunctionName": {
          "Ref": "TestsGetOrderTest88AC940D",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "TestsGetOrderTestServiceRoleDFB27BF0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestsGetOrderTestTrigger2B9067E4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ExecuteOnHandlerChange": true,
        "HandlerArn": {
          "Ref": "TestsGetOrderTestCurrentVersion2705E2D1c5f2213f45730c0d6e64049522f9d2f3",
        },
        "InvocationType": "RequestResponse",
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
        "Timeout": "120000",
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "TestsPostOrderTestCurrentVersion390B109E4024f59419de957764ba65697cc6953e": {
      "Properties": {
        "FunctionName": {
          "Ref": "TestsPostOrderTestEB3C878D",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "TestsPostOrderTestEB3C878D": {
      "DependsOn": [
        "TestsPostOrderTestServiceRoleDefaultPolicyB49CCB30",
        "TestsPostOrderTestServiceRole0D9C5B86",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "my-bucket",
          "S3Key": "my-key",
        },
        "Environment": {
          "Variables": {
            "API_ENDPOINT": "https://api.example.matthewbonig.com/v1/order/",
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Fn::ImportValue": "Api:ExportsOutputRefMyTable794EDED1558A264E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestsPostOrderTestServiceRole0D9C5B86",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestsPostOrderTestServiceRole0D9C5B86": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestsPostOrderTestServiceRoleDefaultPolicyB49CCB30": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:DeleteItem",
                "dynamodb:Query",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestsPostOrderTestServiceRoleDefaultPolicyB49CCB30",
        "Roles": [
          {
            "Ref": "TestsPostOrderTestServiceRole0D9C5B86",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestsPostOrderTestTrigger4038AF83": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ExecuteOnHandlerChange": true,
        "HandlerArn": {
          "Ref": "TestsPostOrderTestCurrentVersion390B109E4024f59419de957764ba65697cc6953e",
        },
        "InvocationType": "RequestResponse",
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
        "Timeout": "120000",
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
